contrib/selinux/syslog_ng.sh

```diff
@@ -125,39 +127,338 @@ setup_vars() {
+			return 0
+		else
+			return 1
+		fi
+	elif syslog_ng_is_ose; then
+		if plugin_version_is_greater_than "${SYSLOG_VERSION}" "${OSE_SUPPORTED_VER}"; then
+			return 0
+		else
+			return 1
+		fi
+	else
+		return 1
+	fi
+}
+
+
+validate_plugin_metadata() {
+	local plugin_dir PLUGIN_NAME PLUGIN_FEATURE PE_SUPPORT_ADDED_VERSION OSE_SUPPORT_ADDED_VERSION
+
+	plugin_dir="${1}"
+	PLUGIN_NAME="${2}"
+	PLUGIN_FEATURE="${3}"
+	PE_SUPPORT_ADDED_VERSION="${4}"
+	OSE_SUPPORT_ADDED_VERSION="${5}"
+
+	if [[ ( -z "${PLUGIN_NAME}" || -z "${PLUGIN_FEATURE}" ) || ( -z "${PE_SUPPORT_ADDED_VERSION}" && -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 243 
+	elif [[ ( "${PLUGIN_FEATURE}" == "PE_ONLY" && -z "${PE_SUPPORT_ADDED_VERSION}" ) || ( "${PLUGIN_FEATURE}" == "OSE_ONLY" && -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 242
+	elif [[ ( "${PLUGIN_FEATURE}" == "ALL" ) && ( -z "${PE_SUPPORT_ADDED_VERSION}" || -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 241
+	fi
+}
+
+
+run_plugin_installers() {
+	local choice
+
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
+		validate_plugin_metadata "${plugin_dir}" "${PLUGIN_NAME}" "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"
+
+		choice=$( echo -ne "${PLUGIN_DB}" | fgrep "${plugin_dir}:" | fetch_plugin_choice )
+		if [ "x${choice}x" == "xYx" ] ; then
+			if plugin_install_precheck "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"; then
+				if already_loaded; then
+					echo "The SELinux module for '${PLUGIN_NAME}' is already loaded! Skipping install..." >&2
+				else
+					"${plugin_dir}/setup.sh" --install-dir "${INSTALL_PATH}" --os "${OS_VERSION}"
+				fi
+			fi
+		fi
+	done < <( dump_PLUGIN_DB | fetch_plugin_dir )
+}
+
+
+run_plugin_callbacks() {
+	local choice buffer
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
+		validate_plugin_metadata "${plugin_dir}" "${PLUGIN_NAME}" "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"
+
+		choice=$( dump_PLUGIN_DB | fgrep "${plugin_dir}:" | fetch_plugin_choice )
+		if [ "x${choice}x" == "xYx" ] ; then
+			for type in te if fc; do
+				buffer=$( return_files "${type}" )
+				[ -n "${buffer}" ] && cat >> "syslog_ng.${type}" <<<"${buffer}"
+			done
+		fi
+	done < <( dump_PLUGIN_DB | fetch_plugin_dir )
+}
+
+
+remove_plugin_modules() {
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
```

Comment:
Do you need to source the include script many times?
```
syslog_ng.sh=query_plugins() {
syslog_ng.sh=run_plugin_installers() {
syslog_ng.sh=run_plugin_callbacks() {
syslog_ng.sh=remove_plugin_modules() {
```
I see that for install/remove you need to do that, but for the other steps it seems unnecessary.

Reply:
Yes, because those files the metadata for the plugins. After reading them and using them for whatever purpose, we discard/overwrite that information when dealing with the next plugin.